src/averageAges.js

```diff
@@ -1,61 +1,56 @@
 'use strict';
 
 /**
- * Implement calculateMenAverageAge function
- *
- * Function returns average age of men in array. If `century` is specified then
- * function calculates average age only for men who died in this century
- *
- * To calculate century:
- * Divide year of person's death by 100: Math.ceil(person.died / 100)
- *
  * @param {object[]} people
  * @param {number} century - optional
+ * @param {boolean} withChildren - optional
+ * @param {boolean} onlyWithSon - optional
  *
  * @return {number}
  */
+
 function calculateMenAverageAge(people, century) {
-  // write code here
-  // learn how to use array methods like .filter .map .some .every .find .reduce
-  // avoid using loop and forEach
-  // replace `if ()` statement with &&, || or ?:
-  // without nesting
+  const men = people.filter(
+    !century
+      ? ({ sex }) => sex === 'm'
+      : ({ sex, died }) => sex === 'm' && Math.ceil(died / 100) === century
+  );
+
+  return averageAge(men);
 }
 
-/**
- * Implement calculateWomenAverageAge function
- *
- * Function returns average age of women in array. If `withChildren` is
- * specified then function calculates average age only for women with children
- *
- * Hint: To check if a woman has children you should find someone who mention
- * her as mother.
- *
- * @param {object[]} people
- * @param {boolean} withChildren - optional
- *
- * @return {number}
- */
 function calculateWomenAverageAge(people, withChildren) {
-  // write code here
+  const women = people.filter(
+    !withChildren
+      ? ({ sex }) => sex === 'f'
+      : ({ name }) =>
+        people.some(({ mother }) => mother === name)
+  );
+
+  return averageAge(women);
 }
 
-/**
- * Implement calculateAverageAgeDiff function.
- *
- * The function returns an average age difference between a child and his or her
- * mother in the array. (A mother's age at child birth)
- *
- * If `onlyWithSon` is specified then function calculates age difference only
- * for sons and their mothers.
- *
- * @param {object[]} people
- * @param {boolean} onlyWithSon - optional
- *
- * @return {number}
- */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // write code here
+  const mothers = people.filter(({ name }) =>
+    people.some(({ mother }) => mother === name));
+
+  const kids = people.filter(
+    !onlyWithSon
+      ? person => people.some(mother =>
+        mother.name === person.mother)
+      : person => people.some((mother, sex) =>
+        mother.name === person.mother && person.sex === 'm')
+  );
+
+  const yearDifference = kids.map(kid =>
+    kid.born - mothers.find((mother) => kid.mother === mother.name).born);
```

Comment:
Cтарайся не розривати колбеки, бо так важко зрозуміти, де закінчується один колбек і починається інший
```suggestion
  const yearDifference = kids.map(
    kid => kid.born - mothers.find(
      (mother) => kid.mother === mother.name
  ).born);
```